### ISSUE TITLE
Triplet loss

### ISSUE THREAD

**Issue by distant1219** [KaiyangZhou/deep-person-reid#19]
I don‘t understand how you generate triplet images to train the model. It seems that you rewrite the sampler class. But why you set the num_instances is 4? Thanks for your answer.

**Comment by ercanburak** [comment]
The triplet mining strategy named "Batch Hard" from the paper "[In Defense of the Triplet Loss for Person Re-Identification](https://arxiv.org/abs/1703.07737)" is employed here. 

In the paper, the batch sampling procedure is explained as follows:
`...the core idea is to form batches by randomly sampling P classes (person identities), and then randomly sampling K images of each class (person), thus resulting in a batch of PK images.`

The sampler class also does the same thing:    
`Randomly sample N identities, then for each identity, randomly sample K instances, therefore batch size is N*K.`

Again, in the paper, the K value is selected as 4 in the experiments. So the sampler class uses this as the default value.

**Comment by distant1219** [comment]
Thanks a lot!

**Comment by distant1219** [comment]
@ercanburak @KaiyangZhou Sorry to bother you again.I have checked the code about triplet loss.I have a question after training market-1501 dataset. I found that only num_identities * num_instances images were used for training at every epoch. For example, in market-1501, we used 751 * 4 = 3004 at each epoch, only 23% of all train images was used. I wonder if it makes the accuracy of the model worse?
